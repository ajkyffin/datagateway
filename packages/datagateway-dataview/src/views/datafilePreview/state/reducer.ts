import DATAFILE_PREVIEWER_DEFAULT from '../defaults';
import {
  DecrementDatafilePreviewerZoomLevelType,
  IncrementDatafilePreviewerZoomLevelType,
  ResetDatafilePreviewerZoomLevelType,
  ToggleDatafilePreviewerDetailsPanePayload,
  ToggleDatafilePreviewerDetailsPaneType,
} from './actions';
import { DGDataViewState } from '../../../state/app.types';

/**
 * Defines the shape of the redux state for {@link DatafilePreviewer}
 */
interface DatafilePreviewerState {
  isDetailsPaneShown: boolean;
  zoomLevel: number;
}

function toggleDatafilePreviewerDetailsPane(
  state: DGDataViewState,
  payload: ToggleDatafilePreviewerDetailsPanePayload
): DGDataViewState {
  return {
    ...state,
    isisDatafilePreviewer: {
      ...state.isisDatafilePreviewer,
      isDetailsPaneShown: payload.shouldShow,
    },
  };
}

function incDatafilePreviewerZoomLevel(
  state: DGDataViewState
): DGDataViewState {
  return {
    ...state,
    isisDatafilePreviewer: {
      ...state.isisDatafilePreviewer,
      zoomLevel:
        state.isisDatafilePreviewer.zoomLevel +
        DATAFILE_PREVIEWER_DEFAULT.zoomLevelStep,
    },
  };
}

function decDatafilePreviewerZoomLevel(
  state: DGDataViewState
): DGDataViewState {
  return {
    ...state,
    isisDatafilePreviewer: {
      ...state.isisDatafilePreviewer,
      // zoom level should not go lower than zoomLevelStep
      // otherwise will result in zoom level <= 0.
      zoomLevel: Math.max(
        DATAFILE_PREVIEWER_DEFAULT.minZoomLevel,
        state.isisDatafilePreviewer.zoomLevel -
          DATAFILE_PREVIEWER_DEFAULT.zoomLevelStep
      ),
    },
  };
}

function resetDatafilePreviewerZoomLevel(
  state: DGDataViewState
): DGDataViewState {
  return {
    ...state,
    isisDatafilePreviewer: {
      ...state.isisDatafilePreviewer,
      zoomLevel: DATAFILE_PREVIEWER_DEFAULT.zoomLevel,
    },
  };
}

const datafilePreviewerInitialState: DatafilePreviewerState = {
  isDetailsPaneShown: DATAFILE_PREVIEWER_DEFAULT.showDetailsPane,
  zoomLevel: DATAFILE_PREVIEWER_DEFAULT.zoomLevel,
};

const datafilePreviewerReducer = {
  [ToggleDatafilePreviewerDetailsPaneType]: toggleDatafilePreviewerDetailsPane,
  [IncrementDatafilePreviewerZoomLevelType]: incDatafilePreviewerZoomLevel,
  [DecrementDatafilePreviewerZoomLevelType]: decDatafilePreviewerZoomLevel,
  [ResetDatafilePreviewerZoomLevelType]: resetDatafilePreviewerZoomLevel,
};

export {
  datafilePreviewerInitialState,
  datafilePreviewerReducer,
  toggleDatafilePreviewerDetailsPane,
  incDatafilePreviewerZoomLevel,
  decDatafilePreviewerZoomLevel,
  resetDatafilePreviewerZoomLevel,
};
export type { DatafilePreviewerState };
