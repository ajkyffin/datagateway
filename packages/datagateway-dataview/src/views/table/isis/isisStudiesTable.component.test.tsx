import React from 'react';
import { createMount } from '@material-ui/core/test-utils';
import ISISStudiesTable from './isisStudiesTable.component';
import { initialState as dgDataViewInitialState } from '../../../state/reducers/dgdataview.reducer';

import { StateType } from '../../../state/app.types';
import {
  Study,
  dGCommonInitialState,
  useStudyCount,
  useStudiesInfinite,
} from 'datagateway-common';
import { ReactWrapper } from 'enzyme';
import configureStore from 'redux-mock-store';
import { QueryClient, QueryClientProvider } from 'react-query';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { Router } from 'react-router';
import { createMemoryHistory, History } from 'history';

jest.mock('datagateway-common', () => {
  const originalModule = jest.requireActual('datagateway-common');

  return {
    __esModule: true,
    ...originalModule,
    useStudyCount: jest.fn(),
    useStudiesInfinite: jest.fn(),
  };
});

describe('ISIS Studies table component', () => {
  let mount;
  let mockStore;
  let state: StateType;
  let rowData: Study[];
  let history: History;

  const createWrapper = (): ReactWrapper => {
    const store = mockStore(state);
    return mount(
      <Provider store={store}>
        <Router history={history}>
          <QueryClientProvider client={new QueryClient()}>
            <ISISStudiesTable instrumentId="1" />
          </QueryClientProvider>
        </Router>
      </Provider>
    );
  };

  beforeEach(() => {
    mount = createMount();
    rowData = rowData = [
      {
        id: 1,
        study: {
          id: 1,
          pid: 'doi',
          name: 'Test 1',
          modTime: '2000-01-01',
          createTime: '2000-01-01',
        },
      },
    ];
    history = createMemoryHistory();

    mockStore = configureStore([thunk]);
    state = JSON.parse(
      JSON.stringify({
        dgdataview: dgDataViewInitialState,
        dgcommon: dGCommonInitialState,
      })
    );

    (useStudyCount as jest.Mock).mockReturnValue({
      data: 1,
      isLoading: false,
    });

    (useStudiesInfinite as jest.Mock).mockReturnValue({
      data: { pages: [rowData] },
      fetchNextPage: jest.fn(),
    });
  });

  afterEach(() => {
    mount.cleanUp();
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const wrapper = createWrapper();
    expect(wrapper.find('VirtualizedTable').props()).toMatchSnapshot();
  });

  it('calls the correct data fetching hooks on load', () => {
    const instrumentId = '1';
    createWrapper();
    expect(useStudyCount).toHaveBeenCalledWith([
      {
        filterType: 'where',
        filterValue: JSON.stringify({
          'studyInvestigations.investigation.investigationInstruments.instrument.id': {
            eq: instrumentId,
          },
        }),
      },
    ]);
    expect(useStudiesInfinite).toHaveBeenCalledWith([
      {
        filterType: 'where',
        filterValue: JSON.stringify({
          'studyInvestigations.investigation.investigationInstruments.instrument.id': {
            eq: instrumentId,
          },
        }),
      },
      {
        filterType: 'include',
        filterValue: JSON.stringify({
          studyInvestigations: 'investigation',
        }),
      },
    ]);
  });

  it('calls useStudiesInfinite when loadMoreRows is called', () => {
    const fetchNextPage = jest.fn();
    (useStudiesInfinite as jest.Mock).mockReturnValueOnce({
      data: { pages: [rowData] },
      fetchNextPage,
    });
    const wrapper = createWrapper();

    wrapper.find('VirtualizedTable').prop('loadMoreRows')({
      startIndex: 50,
      stopIndex: 74,
    });

    expect(fetchNextPage).toHaveBeenCalledWith({
      pageParam: { startIndex: 50, stopIndex: 74 },
    });
  });

  it('updates filter query params on text filter', () => {
    const wrapper = createWrapper();

    const filterInput = wrapper
      .find('[aria-label="Filter by studies.name"] input')
      .first();
    filterInput.instance().value = 'test';
    filterInput.simulate('change');

    expect(history.length).toBe(2);
    expect(history.location.search).toBe(
      `?filters=${encodeURIComponent(
        '{"name":{"value":"test","type":"include"}}'
      )}`
    );
    filterInput.instance().value = '';
    filterInput.simulate('change');

    expect(history.length).toBe(3);
    expect(history.location.search).toBe('?');
  });

  // TODO: Not finding the end date filter (due to the rowData?)
  it.skip('updates filter query params on date filter', () => {
    const wrapper = createWrapper();

    const filterInput = wrapper.find('[aria-label="End Date date filter to"]');
    filterInput.instance().value = '2019-08-06';
    filterInput.simulate('change');

    expect(history.length).toBe(2);
    expect(history.location.search).toBe(
      `?filters=${encodeURIComponent(
        '{"studyInvestigations.investigation.endDate":{"endDate":"2019-08-06"}}'
      )}`
    );

    filterInput.instance().value = '';
    filterInput.simulate('change');

    expect(history.length).toBe(3);
    expect(history.location.search).toBe('?');
  });

  it('updates sort query params on sort', () => {
    const wrapper = createWrapper();

    wrapper
      .find('[role="columnheader"] span[role="button"]')
      .first()
      .simulate('click');

    expect(history.length).toBe(2);
    expect(history.location.search).toBe(
      `?sort=${encodeURIComponent('{"name":"asc"}')}`
    );
  });

  it('renders studies name as a link', () => {
    const wrapper = createWrapper();

    expect(
      wrapper.find('[aria-colindex=1]').find('p').children()
    ).toMatchSnapshot();
  });
});
