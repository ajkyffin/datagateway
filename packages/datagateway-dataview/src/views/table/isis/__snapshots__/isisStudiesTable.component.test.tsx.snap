// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ISIS Studies table component renders correctly 1`] = `
Object {
  "columns": Array [
    Object {
      "cellContentRenderer": [Function],
      "dataKey": "name",
      "filterComponent": [Function],
      "icon": Object {
        "$$typeof": Symbol(react.memo),
        "compare": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      },
      "label": "studies.name",
    },
    Object {
      "cellContentRenderer": [Function],
      "dataKey": "studyInvestigations.investigation.title",
      "filterComponent": [Function],
      "icon": Object {
        "$$typeof": Symbol(react.memo),
        "compare": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      },
      "label": "studies.title",
    },
    Object {
      "cellContentRenderer": [Function],
      "dataKey": "pid",
      "filterComponent": [Function],
      "icon": Object {
        "$$typeof": Symbol(react.memo),
        "compare": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      },
      "label": "studies.pid",
    },
    Object {
      "cellContentRenderer": [Function],
      "dataKey": "studyInvestigations.investigation.startDate",
      "defaultSort": "desc",
      "filterComponent": [Function],
      "icon": Object {
        "$$typeof": Symbol(react.memo),
        "compare": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      },
      "label": "studies.start_date",
    },
    Object {
      "cellContentRenderer": [Function],
      "dataKey": "studyInvestigations.investigation.endDate",
      "filterComponent": [Function],
      "icon": Object {
        "$$typeof": Symbol(react.memo),
        "compare": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      },
      "label": "studies.end_date",
    },
  ],
  "data": Array [
    Object {
      "createTime": "2000-01-01",
      "id": 1,
      "modTime": "2000-01-01",
      "name": "Test 1",
      "pid": "doi",
    },
  ],
  "loadMoreRows": [Function],
  "onSort": [Function],
  "sort": Object {
    "studyInvestigations.investigation.startDate": "desc",
  },
  "totalRowCount": 1,
}
`;

exports[`ISIS Studies table component renders studies name as a link 1`] = `
<ForwardRef(Link)
  component={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "propTypes": Object {
        "innerRef": [Function],
        "onClick": [Function],
        "replace": [Function],
        "target": [Function],
        "to": [Function],
      },
      "render": [Function],
    }
  }
  to="/browseStudyHierarchy/instrument/1/study/1"
>
  <MuiLinkRoot
    className="MuiLink-root MuiLink-underlineAlways"
    color="primary"
    component={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "propTypes": Object {
          "innerRef": [Function],
          "onClick": [Function],
          "replace": [Function],
          "target": [Function],
          "to": [Function],
        },
        "render": [Function],
      }
    }
    onBlur={[Function]}
    onFocus={[Function]}
    ownerState={
      Object {
        "children": "Test 1",
        "color": "primary",
        "component": Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        },
        "data-testid": undefined,
        "focusVisible": false,
        "to": "/browseStudyHierarchy/instrument/1/study/1",
        "underline": "always",
        "variant": "inherit",
      }
    }
    to="/browseStudyHierarchy/instrument/1/study/1"
    variant="inherit"
  >
    <Noop />
    <ForwardRef(Typography)
      className="MuiLink-root MuiLink-underlineAlways css-okiaqn-MuiLink-root"
      color="primary"
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      onBlur={[Function]}
      onFocus={[Function]}
      to="/browseStudyHierarchy/instrument/1/study/1"
      variant="inherit"
    >
      <MuiTypographyRoot
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "propTypes": Object {
              "innerRef": [Function],
              "onClick": [Function],
              "replace": [Function],
              "target": [Function],
              "to": [Function],
            },
            "render": [Function],
          }
        }
        className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-okiaqn-MuiLink-root"
        onBlur={[Function]}
        onFocus={[Function]}
        ownerState={
          Object {
            "align": "inherit",
            "children": "Test 1",
            "className": "MuiLink-root MuiLink-underlineAlways css-okiaqn-MuiLink-root",
            "color": "primary.main",
            "component": Object {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": Object {
                "innerRef": [Function],
                "onClick": [Function],
                "replace": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            },
            "data-testid": undefined,
            "gutterBottom": false,
            "noWrap": false,
            "onBlur": [Function],
            "onFocus": [Function],
            "paragraph": false,
            "sx": Object {
              "color": "primary.main",
            },
            "to": "/browseStudyHierarchy/instrument/1/study/1",
            "variant": "inherit",
            "variantMapping": Object {
              "body1": "p",
              "body2": "p",
              "h1": "h1",
              "h2": "h2",
              "h3": "h3",
              "h4": "h4",
              "h5": "h5",
              "h6": "h6",
              "inherit": "p",
              "subtitle1": "h6",
              "subtitle2": "h6",
            },
          }
        }
        sx={
          Object {
            "color": "primary.main",
          }
        }
        to="/browseStudyHierarchy/instrument/1/study/1"
      >
        <Noop />
        <Link
          className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1iwtne7-MuiTypography-root-MuiLink-root"
          onBlur={[Function]}
          onFocus={[Function]}
          to="/browseStudyHierarchy/instrument/1/study/1"
        >
          <LinkAnchor
            className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1iwtne7-MuiTypography-root-MuiLink-root"
            href="/browseStudyHierarchy/instrument/1/study/1"
            navigate={[Function]}
            onBlur={[Function]}
            onFocus={[Function]}
          >
            <a
              className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1iwtne7-MuiTypography-root-MuiLink-root"
              href="/browseStudyHierarchy/instrument/1/study/1"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              Test 1
            </a>
          </LinkAnchor>
        </Link>
      </MuiTypographyRoot>
    </ForwardRef(Typography)>
  </MuiLinkRoot>
</ForwardRef(Link)>
`;
